tools/src/main/java/org/apache/kafka/tools/TransactionalMessageCopier.java

```diff
@@ -226,80 +259,133 @@ private static String toJsonString(Map<String, Object> data) {
+            System.out.println(shutDownString(totalMessageProcessed.get(),
+                numMessagesProcessedSinceLastRebalance.get(), remainingMessages.get(), transactionalId));
         });
 
+        final boolean useGroupMetadata = parsedArgs.getBoolean("useGroupMetadata");
         try {
             Random random = new Random();
-            while (0 < remainingMessages.get()) {
-                System.out.println(statusAsJson(numMessagesProcessed.get(), remainingMessages.get(), transactionalId));
+            while (remainingMessages.get() > 0) {
+                System.out.println(statusAsJson(totalMessageProcessed.get(),
+                    numMessagesProcessedSinceLastRebalance.get(), remainingMessages.get(), transactionalId, "ProcessLoop"));
                 if (isShuttingDown.get())
                     break;
-                int messagesInCurrentTransaction = 0;
-                long numMessagesForNextTransaction = Math.min(numMessagesPerTransaction, remainingMessages.get());
+                long messagesSentWithinCurrentTxn = 0L;
+                long messagesNeededForCurrentTxn = remainingMessages.get();
 
                 try {
                     producer.beginTransaction();
-                    while (messagesInCurrentTransaction < numMessagesForNextTransaction) {
+                    Map<Integer, Integer> partitionCount = new HashMap<>();
+                    while (messagesSentWithinCurrentTxn < messagesNeededForCurrentTxn) {
                         ConsumerRecords<String, String> records = consumer.poll(Duration.ofMillis(200));
+                        log.info("number of consumer records fetched: {}", records.count());
+                        if (messageCap.get() <= 0) {
+                            // We could see no more message needed for processing after poll
+                            break;
+                        }
                         for (ConsumerRecord<String, String> record : records) {
                             producer.send(producerRecordFromConsumerRecord(outputTopic, record));
-                            messagesInCurrentTransaction++;
+                            messagesSentWithinCurrentTxn++;
+                            partitionCount.put(record.partition(), partitionCount.getOrDefault(record.partition(), 0) + 1);
                         }
+                        messagesNeededForCurrentTxn = Math.min(numMessagesPerTransaction, remainingMessages.get());
+                    }
+
+                    log.info("Messages sent in current transaction: {}", messagesSentWithinCurrentTxn);
+                    for (Map.Entry<Integer, Integer> entry : partitionCount.entrySet()) {
+                        log.info("Partition {} has contributed {} records", entry.getKey(), entry.getValue());
+                    }
+
+                    if (useGroupMetadata) {
+                        producer.sendOffsetsToTransaction(consumerPositions(consumer), consumer.groupMetadata());
```

Comment:
I had a similar question on the other PR: are we suggesting for manual assignment to use the original overload (hence we would never deprecate it)?

Reply:
I think the main reason why we would not deprecate the existing overload is a "producer only" application for which there is not even a consumer.